lib/adapters/show-document-adapter.ts

```diff
@@ -0,0 +1,59 @@
+import { shell } from "electron"
+import { LanguageClientConnection, ShowDocumentParams, ShowDocumentResult } from "../languageclient"
+import { TextEditor } from "atom"
+import Convert from "../convert"
+
+/** Public: Adapts the window/showDocument command to Atom's text editors or external programs. */
+export default class ShowDocumentAdapter {
+  /** {@inheritDoc attach} */
+  public attach(arg: Parameters<typeof attach>[0]): ReturnType<typeof attach> {
+    attach(arg)
+  }
+  /** {@inheritDoc showDocument} */
+  public onShowDocument(...args: Parameters<typeof showDocument>): ReturnType<typeof showDocument> {
+    return showDocument(...args)
+  }
+}
+
+/**
+ * Public: Attach to a {LanguageClientConnection} to recieve requests to show documents.
+ */
+export function attach(connection: LanguageClientConnection): void {
+  connection.onShowDocument(showDocument)
+}
+
+/**
+ * Public: show documents inside Atom text editor or in external programs
+ *
+ * @param params The {ShowDocumentParams} received from the language server
+ *   indicating the document to be displayed as well as other metadata.
+ * @returns {Promise<ShowDocumentResult>} with a `success: boolean` property specifying if the operation was sucessful
+ * {@inheritDoc ShowDocumentParams}
+ */
+export async function showDocument(params: ShowDocumentParams): Promise<ShowDocumentResult> {
+  try {
+    if (!params.external) {
+      // open using atom.workspace
+      const view = await atom.workspace.open(Convert.uriToPath(params.uri), {
+        activateItem: params.takeFocus,
+        activatePane: params.takeFocus,
+        pending: true,
```

Comment:
Is there a reason this is opened as a pending item? (Not saying it shouldn't, just wondering if that is part of the spec?)

Reply:
This is what it means. Opening in a pending is an Atom thing and not part of the spec.
```ts
    /**
     *  A Boolean indicating whether or not the item should be opened in a pending state.
     *  Existing pending items in a pane are replaced with new pending items when they
     *  are opened.
     */
```

Reply:
I know what it is. I'm not sure if that is actually what the user would want. Or if there is some way we could make that configurable.

Reply:
Maybe we should set it as the value of `atom.config.get('core.allowPendingPaneItems')`

Reply:
Actually it looks like Atom ignores it if that setting is set anyway.

https://github.com/atom/atom/blob/757478442696b302bb8f9ad16e93b30f9350d949/src/workspace.js#L1102

Reply:
In practice, I don't see any difference. I am not sure what it exactly does. Maybe better to improve the documentation for the `pending` option before changing the code here.